src/main/java/build/buf/protovalidate/Uri.java

```diff
@@ -0,0 +1,959 @@
+        return true;
+      }
+    }
+
+    this.index = start;
+
+    return false;
+  }
+
+  /**
+   * Reports whether string contains a valid authority.
+   *
+   * <p>Method parses the rule:
+   *
+   * <pre>authority = [ userinfo "@" ] host [ ":" port ]
+   *
+   * Lead by double slash ("") and terminated by "/", "?", "#", or end of URI.
+   */
+  private boolean authority() {
+    int start = this.index;
+
+    if (this.userinfo()) {
+      if (!this.take('@')) {
+        this.index = start;
+        return false;
+      }
+    }
+
+    if (!this.host()) {
+      this.index = start;
+      return false;
+    }
+
+    if (this.take(':')) {
+      if (!this.port()) {
+        this.index = start;
+        return false;
+      }
+    }
+
+    if (!this.isAuthorityEnd()) {
+      this.index = start;
+      return false;
+    }
+
+    return true;
+  }
+
+  /**
+   * Reports whether the current position is the end of the authority.
+   *
+   * <p>The authority component [...] is terminated by one of the following:
+   *
+   * <ul>
+   *   <li>the next slash ("/")
+   *   <li>question mark ("?")
+   *   <li>number sign ("#") character
+   *   <li>the end of the URI.
+   * </ul>
+   */
+  private boolean isAuthorityEnd() {
+    if (this.index >= this.str.length()) {
+      return true;
+    }
+    char c = this.str.charAt(this.index);
+    return (c == '?' || c == '#' || c == '/');
+  }
+
+  /**
+   * Reports whether string contains a valid userinfo.
+   *
+   * <p>Method parses the rule:
+   *
+   * <pre>userinfo = *( unreserved / pct-encoded / sub-delims / ":" )
+   *
+   * Terminated by "@" in authority.
+   */
+  private boolean userinfo() {
+    int start = this.index;
+
+    while (true) {
+      if (this.unreserved() || this.pctEncoded() || this.subDelims() || this.take(':')) {
+        continue;
+      }
+
+      if (this.index < this.str.length()) {
+        if (this.str.charAt(this.index) == '@') {
+          return true;
+        }
+      }
+
+      this.index = start;
+
+      return false;
+    }
+  }
+
+  @FunctionalInterface
+  interface UnhexOperation {
+    int unhex(char c);
+  }
+
+  /**
+   * Verifies that str is correctly percent-encoded.
+   *
+   * <p>Note that we essentially want to mimic the behavior of decodeURIComponent, which would fail
+   * on malformed URLs. Java does have various methods for decoding URLs, but none behave
+   * consistently with decodeURIComponent.
+   *
+   * <p>The code below is a combination of `checkHostPctEncoded` from the protovalidate-go
+   * implementation and Java's java.net.URI#decode methods.
+   */
+  private boolean checkHostPctEncoded(String str) {
+    UnhexOperation fn =
+        c -> {
+          if ('0' <= c && c <= '9') {
+            return c - '0';
+          } else if ('a' <= c && c <= 'f') {
+            return c - 'a' + 10;
+          } else if ('A' <= c && c <= 'F') {
+            return c - 'A' + 10;
+          }
+
+          return 0;
+        };
```

Comment:
IMO a `private static unhex` is preferable over the lambda here.

Reply:
Fixed. Should I fix the one in Go also?